CHANGELOG.md

```diff
@@ -4,8 +4,9 @@ dev
  - Fix bug when attempting to parse function with a magic scope parameter type hint (eg `function (self $param) { ... }`.
  - Implement `Analysis\TolerantExpressionAnalyser` which will convert analysis exceptions into the *mixed* type.
  - Add `Providers\DSL\Compilation\Parameters\ParameterCollection::contains` to check whether the collection contains a parameter.
- - Add `Providers\DSL\Compilation\Parameters\ParameterCollection::remove` to remove a previously added parameter.\
+ - Add `Providers\DSL\Compilation\Parameters\ParameterCollection::remove` to remove a previously added parameter.
  - Introduce `Analysis\INativeType::TYPE_NUMERIC` acting as a union type for `Analysis\INativeType::TYPE_INT` and `Analysis\INativeType::TYPE_DOUBLE`  
+ - Change identity hash to hash instances of `DateTimeInterface` by class and timestamp hence treating it as a value type throughout the PINQ API
```

Comment:
@Ocramius Yep makes sense, like above an unopinionated approach would be the `IClassIdentifier` (user provides implementations for their classes) that would modify how comparisons are made between classes, as the user chooses. The problem is that then the API and how it behaves is moving target, to the whim of the user. Query providers (other implementations of the API) would most likely am at the largest possible support for the API but this becomes harder as these core behaviours are no longer well-defined. 

So really it comes down to priorities and the formalization of behaviour. Define a set standards of value of objects such as `DateTime` (opinionated) or let the user define such behaviour and instead uery providers who implement the API now behave different from the customizations from the user. As usual both sides have there pros and cons :/